sycl/include/sycl/ext/oneapi/properties/properties.hpp

```diff
@@ -17,6 +26,193 @@
+template <> struct properties_sorter<std::integer_sequence<int>> {
+  using type = properties_type_list<>;
+};
+
+template <int... IdxSeq, typename... property_tys>
+struct properties_sorter<std::integer_sequence<int, IdxSeq...>,
+                         property_tys...> {
+  static constexpr auto sorted_indices = []() constexpr {
+    int idx = 0;
+    int N = sizeof...(property_tys);
+    // TODO: Use C++20 constexpr std::sort if available.
+    std::array to_sort{std::pair{property_tys::sort_key, idx++}...};
+    auto swap_pair = [](auto &x, auto &y) constexpr {
+      auto tmp_first = x.first;
+      auto tmp_second = x.second;
+      x.first = y.first;
+      x.second = y.second;
+      y.first = tmp_first;
+      y.second = tmp_second;
+    };
+    for (int i = 0; i < N; ++i)
+      for (int j = i; j < N; ++j)
+        if (to_sort[j].first < to_sort[i].first)
+          swap_pair(to_sort[i], to_sort[j]);
+
+    std::array<int, sizeof...(property_tys)> sorted_indices{};
+    for (int i = 0; i < N; ++i)
+      sorted_indices[i] = to_sort[i].second;
+
+    return sorted_indices;
+  }();
+
+  using type = properties_type_list<
+      nth_type_t<sorted_indices[IdxSeq], property_tys...>...>;
+};
+
+struct property_key_tag_base {};
+
+template <typename property_key_t>
+struct property_key_tag : property_key_tag_base {};
+
+template <typename property_t, typename property_key_t = property_t>
+struct property_base : property_key_tag<property_key_t> {
+protected:
+  using key_t = property_key_t;
+  constexpr property_t get_property(property_key_tag<key_t>) const {
+    return *static_cast<const property_t *>(this);
+  }
+};
+} // namespace detail
+
+using property_sort_key_t = std::string_view;
+
+template <typename properties_type_list_ty> class properties;
+
+template <typename T> struct is_property_list : std::false_type {};
+template <typename PropListTy>
+struct is_property_list<properties<PropListTy>> : std::true_type {};
+template <typename T>
+inline constexpr bool is_property_list_v = is_property_list<T>::value;
+
+template <typename T>
+inline constexpr bool is_property_v =
+    std::is_base_of_v<detail::property_key_tag_base, T> &&
+    !is_property_list_v<T>;
+
+template <typename... property_tys>
+class __SYCL_EBO properties<detail::properties_type_list<property_tys...>>
+    : public property_tys... {
+  static_assert((is_property_v<property_tys> && ...));
+  static_assert(
+      []() constexpr {
+        if constexpr (sizeof...(property_tys) == 0) {
+          return true;
+        } else {
+          const std::array sort_keys = {property_tys::sort_key...};
+          // std::is_sorted isn't constexpr until C++20.
+          if (sort_keys.empty())
+            return true;
+          for (std::size_t idx = 1; idx < sort_keys.size(); ++idx)
+            if (sort_keys[idx - 1] >= sort_keys[idx])
+              return false;
+          return true;
+        }
+      }(),
+      "Properties must be sorted and non-repeating!");
```

Comment:
Agree with negative unit test. Not sure about repeating sort_key. I assume you're talking about the case when two distinct unrelated properties happened to have the same sort key by mistake. However, that is undistinguishable from a valid scenario when compile-time properties that carry extra information other than their mere presence do share the sort_key. For example, 

```
enum class data_placement { blocked, striped };
template <data_placement placement>
struct data_placement_property
    : detail::property_base<data_placement_property<placement>,
                            struct data_placement_property_key> {
  static constexpr mock_property_sort_key_t sort_key = 3;
  static constexpr bool is_blocked() {
    return placement == data_placement::blocked;
  }
};
```

We don't want to allow `properties{data_placement_property<blocked>{}, data_placement_property<striped>{}}` and we catch that because they have the same sort key, but the fact that they do have it is by design and is intentional.

Reply:
> I had to add a special base case

Thanks! I wouldn't have guessed how to make that work on my own...

Reply:
We can distinguish `properties{data_placement_property<blocked>{}, data_placement_property<striped>{}}` from `properties{prop_a{}, prop_b{}}` where propb_a and prob_b have the same sort-key by mistake. While in both cases the property type is different and the sort key is the same, the first case has the same property key type and the 2nd has two different property key types. This difference allows us to issue different diagnostic.

Reply:
I think latest revision addresses this thread, please "resolve conversation" if you agree.